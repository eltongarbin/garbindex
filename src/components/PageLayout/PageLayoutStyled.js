import styled from 'styled-components';
import { Typography, IconButton } from '@material-ui/core';

export const RootContainer = styled.div`
  flex-grow: 1;
`;

export const HeaderTitle = styled(Typography)`
  flex-grow: 1;
`;

export const MenuButton = styled(IconButton)`
  && {
    margin-left: -12px;
    margin-right: 4px;
  }
`;
