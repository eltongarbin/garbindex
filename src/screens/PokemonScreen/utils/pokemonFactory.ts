const defaultPokemon = {
  height: 6,
  id: 35,
  name: 'clefairy',
  sprites: {
    front_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/35.png'
  },
  stats: [
    {
      base_stat: 35,
      name: 'speed',
      id: 6
    },
    {
      base_stat: 65,
      name: 'special-defense',
      id: 5
    },
    {
      base_stat: 60,
      name: 'special-attack',
      id: 4
    },
    {
      base_stat: 48,
      name: 'defense',
      id: 3
    },
    {
      base_stat: 45,
      name: 'attack',
      id: 2
    },
    {
      base_stat: 70,
      name: 'hp',
      id: 1
    }
  ],
  weight: 75,
  image:
    'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/35.png',
  abilitiesById: {
    '56': {
      id: 56,
      name: 'cute-charm'
    },
    '98': {
      id: 98,
      name: 'magic-guard'
    },
    '132': {
      id: 132,
      name: 'friend-guard'
    }
  },
  typesById: {
    '18': {
      id: 18,
      name: 'fairy'
    }
  },
  evolvedFrom: 'cleffa'
};

export function create(overrides: any = defaultPokemon) {
  return { ...defaultPokemon, ...overrides };
}
