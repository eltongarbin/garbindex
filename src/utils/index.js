export { default as history } from './history';
export { default as theme } from './theme';
