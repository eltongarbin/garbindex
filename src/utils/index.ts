export { default as history } from './history';
export { default as theme } from './theme';

export const extractParamId = (url: any) => {
  const [id] = url.match(/(\d+)(?!.*\d)/);
  return id;
};
