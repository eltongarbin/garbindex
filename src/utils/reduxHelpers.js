import { createAction } from 'redux-actions';
import { call, put } from 'redux-saga/effects';
import isEmpty from 'lodash.isempty';

export const createType = (stateKey, type) =>
  `${process.env.REACT_APP_NAME}/${stateKey}/${type}`;

export const createAsyncTypes = (stateKey, type) => ({
  REQUEST: `${createType(stateKey, type)}_REQUEST`,
  SUCCESS: `${createType(stateKey, type)}_SUCCESS`,
  FAILURE: `${createType(stateKey, type)}_FAILURE`
});

export const createAsyncActions = (asyncTypes) => ({
  request: createAction(asyncTypes.REQUEST),
  receive: createAction(asyncTypes.SUCCESS),
  error: createAction(asyncTypes.FAILURE)
});

export function* genericAsyncResolver(
  fnService,
  parameter,
  asyncActions,
  cbFormatData
) {
  try {
    const data = yield call(fnService, parameter);

    let dataFormatted;
    if (cbFormatData && !isEmpty(data)) {
      dataFormatted = yield cbFormatData(data);
    }

    yield put(asyncActions.receive(dataFormatted || data));
  } catch (error) {
    yield put(asyncActions.error(error));
  }
}
