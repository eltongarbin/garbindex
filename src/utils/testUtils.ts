/* eslint-disable react/forbid-foreign-prop-types */
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, conformingProps) => {
  const propError = checkPropTypes(
    component.propTypes,
    conformingProps,
    'prop',
    component.name
  );

  expect(propError).toBeUndefined();
};
