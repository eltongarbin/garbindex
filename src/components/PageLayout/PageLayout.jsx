import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Toolbar, AppBar, Grid } from '@material-ui/core';
import { ArrowBack as ArrowBackIcon } from '@material-ui/icons';

import { RootContainer, HeaderTitle, MenuButton } from './PageLayoutStyled';

function PageLayout({ title, backTo, children }) {
  return (
    <RootContainer>
      <AppBar position="static">
        <Toolbar>
          {backTo && (
            <MenuButton
              color="inherit"
              aria-label="Back"
              component={Link}
              to={backTo}
            >
              <ArrowBackIcon />
            </MenuButton>
          )}
          <HeaderTitle variant="h6" color="inherit">
            {title}
          </HeaderTitle>
        </Toolbar>
      </AppBar>
      <Grid container justify="center" style={{ padding: 16 }} spacing={0}>
        <Grid item xs={12} sm={8} md={6}>
          {children}
        </Grid>
      </Grid>
    </RootContainer>
  );
}

PageLayout.propTypes = {
  title: PropTypes.string.isRequired,
  backTo: PropTypes.string,
  children: PropTypes.any.isRequired
};

export default PageLayout;
