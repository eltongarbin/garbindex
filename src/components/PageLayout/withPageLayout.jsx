import React from 'react';

import PageLayout from './PageLayout';

const withPageLayout = (layoutProps) => (WrappedComponent) => {
  const PageLayoutWrapper = (props) => (
    <PageLayout {...layoutProps}>
      <WrappedComponent {...props} />
    </PageLayout>
  );

  return PageLayoutWrapper;
};

export default withPageLayout;
