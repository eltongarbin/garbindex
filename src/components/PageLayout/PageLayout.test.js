import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { checkProps } from 'utils/testUtils';
import PageLayout from './PageLayout';
import { MenuButton } from './PageLayoutStyled';

describe('<PageLayout />', () => {
  it('matches the snapshot', () => {
    const tree = shallow(
      <PageLayout title="title">
        <div>content</div>
      </PageLayout>
    );

    expect(toJson(tree)).toMatchSnapshot();
  });

  it('does not throw warning with expected props', () => {
    const expectedProps = { title: 'title', children: 'children' };
    checkProps(PageLayout, expectedProps);
  });

  it('should render back button when pass `onBackClick` prop', () => {
    const wrapper = shallow(
      <PageLayout title="title" onBackClick={() => {}}>
        <div>content</div>
      </PageLayout>
    );

    expect(wrapper.find(MenuButton)).toHaveLength(1);
  });

  it('should not render back button when not pass `onBackClick` prop', () => {
    const wrapper = shallow(
      <PageLayout title="title">
        <div>content</div>
      </PageLayout>
    );

    expect(wrapper.find(MenuButton)).toHaveLength(0);
  });

  it('calls `onBackClick` prop upon button click', () => {
    const buttonActionMock = jest.fn();
    const wrapper = shallow(
      <PageLayout title="title" onBackClick={buttonActionMock}>
        <div>content</div>
      </PageLayout>
    );

    wrapper.find(MenuButton).simulate('click');

    expect(buttonActionMock.mock.calls.length).toBe(1);
  });
});
