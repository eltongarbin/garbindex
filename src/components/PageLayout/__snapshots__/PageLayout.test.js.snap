// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageLayout /> matches the snapshot 1`] = `
<styled.div>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Toolbar)>
      <Styled(WithStyles(Typography))
        color="inherit"
        variant="h6"
      >
        title
      </Styled(WithStyles(Typography))>
    </WithStyles(Toolbar)>
  </WithStyles(AppBar)>
  <Styled(WithStyles(Grid))
    container={true}
    justify="center"
    spacing={0}
  >
    <WithStyles(Grid)
      item={true}
      md={6}
      sm={8}
      xs={12}
    >
      <div>
        content
      </div>
    </WithStyles(Grid)>
  </Styled(WithStyles(Grid))>
</styled.div>
`;
